File: akka/shopping-cart-service-scala/src/test/scala/shopping/cart/CreateTableTestUtils.scala
package shopping.cart

import akka.actor.typed.ActorSystem
import akka.stream.alpakka.cassandra.scaladsl.CassandraSessionRegistry
import shopping.cart.ItemPopularityRepositoryImpl.popularityTable

object CreateTableTestUtils {

  def createTables(system: ActorSystem[_]): Unit = {
    import org.slf4j.LoggerFactory
    import akka.projection.cassandra.scaladsl.CassandraProjection
    import scala.concurrent.Await
    import scala.concurrent.duration._

    // ok to block here, main thread
    Await.result(
      CassandraProjection.createOffsetTableIfNotExists()(system),
      30.seconds)

    // use same keyspace for the item_popularity table as the offset store
    val keyspace = system.settings.config
      .getString("akka.projection.cassandra.offset-store.keyspace")
    val session =
      CassandraSessionRegistry(system).sessionFor("akka.persistence.cassandra")

    Await.result(
      session.executeDDL(
        s"""CREATE TABLE IF NOT EXISTS $keyspace.$popularityTable (
      item_id text,
      count counter,
      PRIMARY KEY (item_id))
      """),
      30.seconds)

    LoggerFactory
      .getLogger("shopping.cart.CreateTableTestUtils")
      .info("Created keyspace [{}] and tables", keyspace)
  }
}

File: akka/shopping-cart-service-java/src/test/java/shopping/cart/CreateTableTestUtils.java
package shopping.cart;

import static java.util.concurrent.TimeUnit.SECONDS;

import akka.actor.typed.ActorSystem;
import akka.projection.cassandra.javadsl.CassandraProjection;
import akka.stream.alpakka.cassandra.javadsl.CassandraSession;
import akka.stream.alpakka.cassandra.javadsl.CassandraSessionRegistry;
import org.slf4j.LoggerFactory;

public class CreateTableTestUtils {
  public static void createTables(ActorSystem<?> system) throws Exception {
    // ok to block here, main thread
    CassandraProjection.createOffsetTableIfNotExists(system).toCompletableFuture().get(30, SECONDS);

    // use same keyspace for the item_popularity table as the offset store
    String keyspace =
        system.settings().config().getString("akka.projection.cassandra.offset-store.keyspace");
    CassandraSession session =
        CassandraSessionRegistry.get(system).sessionFor("akka.persistence.cassandra");

    session
        .executeDDL(
            "CREATE TABLE IF NOT EXISTS "
                + keyspace
                + "."
                + ItemPopularityRepositoryImpl.POPULARITY_TABLE
                + " (\n"
                + "item_id text,\n"
                + "count counter,\n"
                + "PRIMARY KEY (item_id))")
        .toCompletableFuture()
        .get(30, SECONDS);

    LoggerFactory.getLogger(CreateTableTestUtils.class)
        .info("Created keyspace [{}] and tables", keyspace);
  }
}
